src/oidc/index.ts

```diff
@@ -48,7 +53,76 @@ export const requestOidcAuthentication = async (
     redirect_uri: string,
     post_logout_redirect_uri: string
 ) => {
-    const client_id = app_id || localStorage.getItem('config.app_id') || '';
+
+    try {
+        const userManager = await createUserManager(app_id, redirect_uri, post_logout_redirect_uri);
+
+        await userManager.signinRedirect();
+        return { userManager };
+    } catch (error) {
+        console.error('Authentication failed:', error);
+        throw error;
+    }
+};
+
+/**
+ * Requests access tokens from the Hydra authorization server. 
+ * The returned access tokens will be used to fetch the original tokens that can be passed to the `authorize` endpoint
+ * @param {string} app_id The app ID of the platform requesting the access tokens
+ * @param {string} redirect_uri The URL to redirect to after authentication. This defaults to the current URL where this function is called
+ * @param {string} post_logout_redirect_uri The URL to redirect to after logout. This defaults to the current URL where this function is called
+ */
+export const requestOidcToken = async (app_id: string, redirect_uri: string = window.location.href, post_logout_redirect_uri: string = window.location.href) => {
+    try {
+        const userManager = await createUserManager(app_id, redirect_uri, post_logout_redirect_uri)
+
+        const user = await userManager.signinCallback();
+
+        return {
+            accessToken: user?.access_token
+        }
+    } catch (err) {
+        console.error("unable to request access tokens: ", err)
+        throw err
+    }
+}
+
+/**
+ * Fetches the tokens that will be passed to the `authorize` endpoint.
+ * 
+ * @param {string} accessToken The access token received after calling `requestOidcToken` successfully
+ */
+export const requestLegacyToken = async (accessToken: string) => {
+    const server_url_from_local_storage = localStorage.getItem('config.server_url') || 'oauth.deriv.com';
+
+    try {
+        const response = await fetch(`https://${server_url_from_local_storage}/oauth2/legacy/tokens`, { 
+            method : 'POST',
+            headers: {
+                'Authorization': `Bearer ${accessToken}`,
+            },
+        });
+        const data = await response.json()
+
+        return data
+    } catch (err) {
+        console.error("unable to request legacy tokens: ", err)
+        throw err
```

Comment:
[M] throw new Error(), and is it possible to give a auth failure code to differentiate steps of failure?

Reply:
I'm creating error types, it will be on a separate PR